### PR TITLE
This change aims to make the test harness more generic and enforces stronger linting rules during test runs.

### PR DESCRIPTION
**Pull request by nmiodice**
More specifically, the following has been done:

- [x] Fail build if any *.go files for changed modules/templates do not
   follow the linting standards set by `go fmt`
- [x] Fail build if any terraform files for changed modules/templates do not
  follow the linting standards set by `terraform fmt`
- [x] Add unit test fixture code that vastly simplifies writing unit tests
  for terraform files
- [x] Fix builds to support fork PRs

This change sets a pattern for generalizing integration tests, but it
does not implement this pattern.

## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [YES] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## What is the current behavior?
-------------------------------------
Unit tests are hard to write

## What is the new behavior?
-------------------------------------
Unit tests are easy to write

## Does this introduce a breaking change?
-------------------------------------
- NO

## Any relevant logs, error output, etc?
-------------------------------------
Sample Test:
```go
package test
import (
    "fmt"
    "os"
    "testing"
    "github.com/gruntwork-io/terratest/modules/random"
    "github.com/gruntwork-io/terratest/modules/terraform"
    "github.com/microsoft/cobalt/test-harness/infratests"
)
var prefix = fmt.Sprintf("cobalt-%s", random.UniqueId())
var datacenter = os.Getenv("DATACENTER_LOCATION")
var tf_options = &terraform.Options{
    TerraformDir: "../../",
    Upgrade:      true,
    Vars: map[string]interface{}{
        "prefix":   prefix,
        "location": datacenter,
    },
    BackendConfig: map[string]interface{}{
        "storage_account_name": os.Getenv("TF_VAR_remote_state_account"),
        "container_name":       os.Getenv("TF_VAR_remote_state_container"),
    },
}
func TestAzureSimple(t *testing.T) {
    test_fixture := infratests.UnitTestFixture{
        GoTest:                t,
        TfOptions:             tf_options,
        ExpectedResourceCount: 3,
        PlanAssertions:        nil,
        ExpectedResourceAttributeValues: infratests.ResourceAttributeValueMapping{
            "azurerm_app_service.main": map[string]string{
                "resource_group_name":            prefix,
                "location":                       datacenter,
                "site_config.0.linux_fx_version": "DOCKER|appsvcsample/static-site:latest",
            },
            "azurerm_app_service_plan.main": map[string]string{
                "kind":       "Linux",
                "location":   datacenter,
                "reserved":   "true",
                "sku.0.size": "S1",
                "sku.0.tier": "Standard",
            },
            "azurerm_resource_group.main": map[string]string{
                "location": datacenter,
                "name":     prefix,
            },
        },
    }
    infratests.RunUnitTests(&test_fixture)
}
```

## Other information
-------------------------------------
N/A

### CODE REVIEW

File: test-harness/docker/base-images/Dockerfile [hunk 1]
```diff
@@ -46,18 +46,20 @@ RUN wget --quiet https://releases.hashicorp.com/terraform/${TF_VERSION}/terrafor
...
 RUN mkdir infra build
 ADD infra/ ./infra
+ADD test-harness/infratests/ ./test-harness/infratests
 
 RUN apt-get purge -y wget build-essential wget unzip curl \
     apt-transport-https lsb-release gpg
 
 # Pull down golang vendor dependenies
 RUN ["dep", "init", "-v"]
+RUN ["dep", "ensure", "-v"]
```
Comment (nmiodice) [inline test-harness/docker/base-images/Dockerfile]: TODO: omit duplicate

File: test-harness/infratests/infratests.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	terraform.WorkspaceSelectOrNew(fixture.GoTest, fixture.TfOptions, workspace_name)
+	defer terraform.RunTerraformCommand(fixture.GoTest, fixture.TfOptions, "workspace", "delete", workspace_name)
+	defer terraform.WorkspaceSelectOrNew(fixture.GoTest, fixture.TfOptions, "default")
+
+	tf_plan_file_path := "out.plan"
+	terraform.RunTerraformCommand(
+		fixture.GoTest,
+		fixture.TfOptions,
+		terraform.FormatArgs(fixture.TfOptions, "plan", "-input=false", "-out", tf_plan_file_path)...)
+	defer os.Remove(tf_plan_file_path)
```
Comment (erikschlegel) [inline test-harness/infratests/infratests.go]: Do you foresee potential naming conflicts with multiple unit test classes for the same template considering that we're writing out a general `out.plan`?

Comment (nmiodice) [reply test-harness/infratests/infratests.go]: If the tests are palatalized then yes. Good catch, I'll make the filename random.

File: test-harness/infratests/infratests.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+// Executes terraform lifecycle events and verifies the correctness of the resulting terraform.
+// The following actions are coordinated:
+//	- Run `terraform init`
+//	- Create new terraform workspace. This helps prevent accidentally deleting resources
+//	- Run `terraform plan`
+//	- Validate terraform plan file.
+func RunUnitTests(fixture *UnitTestFixture) {
+	terraform.Init(fixture.GoTest, fixture.TfOptions)
+
+	workspace_name := random.UniqueId()
```
[the remark this answers is not in the document]
Comment (nmiodice) [reply test-harness/infratests/infratests.go]: Syncing w/ them on Teams. Will not block on this comment.

**Pull request merged**